### PR TITLE
fix(cairo): clean scoreboard on exclusive open to fix txn-txn-infligh…

### PR DESCRIPTION
**Pull request by ideoma**
…t error

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 7 / 8 (87.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 11 / 13 (84.62%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|10|12|83.33%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 13 / 14 (92.86%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|12|13|92.31%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 12 / 15 (80.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|10|13|76.92%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 13 / 28 (46.43%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|11|26|42.31%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 28 / 33 (84.85%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|11|15|73.33%|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|6|7|85.71%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 12 / 13 (92.31%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 12 / 13 (92.31%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 0 / 8 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|0|5|00.00%|
|:large_blue_circle:|io/questdb/std/Files.java|0|1|00.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 8 / 8 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 8 / 8 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 8 / 8 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|

File: core/src/main/c/share/files.c [hunk 1]
```diff
@@ -325,6 +325,52 @@ JNIEXPORT jint JNICALL Java_io_questdb_std_Files_lock
...
+                if (Java_io_questdb_std_Files_allocate(e, cl, fd, size) == JNI_TRUE) {
+                    // downgrade to shared lock
+                    if (flock((int) fd, LOCK_SH) == 0) {
+                        // success
+                        return fd;
+                    }
+                }
+            }
+        } else {
+            if (fileSize >= size || Java_io_questdb_std_Files_allocate(e, cl, fd, size) == JNI_TRUE) {
```
Comment (bluestreak01) [inline core/src/main/c/share/files.c:353]: We should perhaps be more expressive when file size is below the expected value. If this happens errorno is not set but -1 is returned. Would be good to see a test that produces actionable message

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/c/share/files.c (hunk 1 above)
Comment (ideoma) [reply core/src/main/c/share/files.c:353]: Errno will be set in Java_io_questdb_std_Files_allocate if it cannot resize the file as far as I understand. If may look confusing but it means that if file size is greater or resize is successful return OK. So resize will be attempted when file size is below and if it fails flow falls down to unhappy path e.g., close the file and return -1.

File: core/src/main/c/share/files.c [hunk 2]
```diff
@@ -325,6 +325,52 @@ JNIEXPORT jint JNICALL Java_io_questdb_std_Files_lock
...
+    } else {
+        // file size is already 0, no cleanup but allocate the file.
+        if (Java_io_questdb_std_Files_allocate(e, cl, fd, size) == JNI_TRUE && flock((int) fd, LOCK_SH) == 0) {
+            // success
+            return fd;
+        }
+    }
+
+    // Any non-happy path comes here.
+    close(fd);
```
Comment (bluestreak01) [inline core/src/main/c/share/files.c:370]: this will hide errno from `flock` and `ftruncate`

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**